### PR TITLE
UsersSettings RFC: Code to Sync User Settings updates to new tables

### PR DESCRIPTION
**Pull request by msarit**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [X] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
- This PR accomplishes Action Step 3 of the [UsersSettings RFC](https://github.com/forem/rfcs/pull/32#issuecomment-820442032).
- For a newly created user, the code creates one Users::Setting record and one Users::NotificationSetting for that user, and then populates the relevant fields in these records with the User table values for that user.
- When a user is updated, the code populates that user's existing `Users::Setting` or `Users::NotificationSetting` record.

## Related Tickets & Documents
Data Update Script PR: https://github.com/forem/forem/pull/13390

https://github.com/forem/rfcs/pull/32

## QA Instructions, Screenshots, Recordings

1. On the front-end, create a new user (do not use the console, so all relevant callbacks can fire).
2. Thereafter, run the command `Users::Setting.find_by(user_id: id-of-created-user)`. You should get a Users::Setting record back. Check that the attributes in this record match those for the user (eg `User.find(id-of-created-user).attribute_name`).
4. Go through the same flow, but this time update a relevant field for an existing user. That attribute's value should also be updated in the `Users::Setting` record for that user.
5. Repeat the above steps for the user's `Users::NotificationSetting` record.

### UI accessibility concerns?
None. This is all backend work.

## Added tests?

- [X] Yes
- [ ] No, not yet
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

- [ ] I've updated the [Developer Docs](https://docs.forem.com) and/or
      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/), or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] I've updated the README or added inline documentation
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [X] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] What gif best describes this PR or how it makes you feel?

Pooh Bear Stretching

### CODE REVIEW

File: db/schema.rb [hunk 1]
```diff
@@ -1452,7 +1452,7 @@
     t.datetime "updated_at", precision: 6, null: false
     t.bigint "user_id", null: false
     t.boolean "welcome_notifications", default: true, null: false
-    t.index ["user_id"], name: "index_users_notification_settings_on_user_id"
+    t.index ["user_id"], name: "index_users_notification_settings_on_user_id", unique: true
```
Comment (citizen428) [inline db/schema.rb:1455]: Are these schema changes supposed to be part of this PR?

File: app/models/user.rb [hunk 2]
```diff
@@ -575,6 +623,74 @@ def flipper_id
 
   private
 
+  # rubocop:disable Metrics/BlockLength
+  def migrate_users_fields_to_users_settings(users_setting_record)
```
Comment (Ridhwana) [inline app/models/user.rb:627]: @msarit this method does not look like it uses the newly updated user record values (obtained from self in the after_commithook) to update the user settings table?

File: app/models/user.rb [hunk 3]
```diff
@@ -303,6 +346,11 @@ class User < ApplicationRecord
   after_save :subscribe_to_mailchimp_newsletter
 
   after_create_commit :send_welcome_notification
+
+  # NOTE: @msarit `sync_users_settings_table` blows up when trying to sync the
+  # User.profile fields, probably because the user's profile hasn't been created yet
```
Comment (Ridhwana) [inline app/models/user.rb:351]: We'll most likely tackle the profile fields in the ProfileMigration concern above

File: app/models/user.rb (hunk 2 above)
Comment (Ridhwana) [inline app/models/user.rb:627]: We can really simplify this function by generalizing the code a little.We'll tackle it when we pair today.

File: app/models/user.rb (hunk 3 above)
Comment (citizen428) [reply app/models/user.rb:351]: > probably because the user's profile hasn't been created yet

Profiles should get created when the user gets created, so hopefully that's not it.

```ruby
# All new users should automatically have a profile
after_create_commit -> { Profile.create(user: self) }, unless: :_skip_creating_profile
```

File: app/models/user.rb [hunk 4]
```diff
@@ -56,7 +56,7 @@ class User < ApplicationRecord
       attr_accessor :_skip_creating_profile
 
       # All new users should automatically have a profile
-      after_create_commit -> { Profile.create(user: self) }, unless: :_skip_creating_profile
+      after_create_commit :create_user_profile, unless: :_skip_creating_profile
```
Comment (citizen428) [inline app/models/user.rb:59]: Is there a specific reason for turning this lambda into a method? I wanted to have all temporary profile-related code within this inline concern but apart from that, they are functionally equivalent.

File: app/models/user.rb [hunk 5]
```diff
@@ -623,63 +628,30 @@ def flipper_id
...
+        users_setting_record.update(field => field_enums[public_send(field).to_sym])
       else
         users_setting_record.update(field => public_send(field))
       end
     end
 
-    PROFILE_FIELDS_TO_MIGRATE_TO_USERS_SETTINGS_TABLE.each do |field|
-      users_setting_record.update(field => profile.public_send(field)) if profile.public_send(field).present?
-    end
+    sync_relevant_profile_fields_to_user_settings_table(users_setting_record) if profile
```
Comment (citizen428) [inline app/models/user.rb:647]: I'm conflicted about this line: while I understand that we don't want to raise exceptions while syncing I also don't think we should silently ignore the fact that a user is missing their profile because that shouldn't happen in the first place.

File: app/models/user.rb (hunk 4 above)
Comment (rhymes) [reply app/models/user.rb:59]: I would revert this change as well, given that it's part of the included concern and not strictly part of the `User` model

File: app/models/user.rb [hunk 6]
```diff
@@ -575,6 +628,45 @@ def flipper_id
...
+        users_setting_record.update(field => public_send(field))
+      end
+    end
+
+    sync_relevant_profile_fields_to_user_settings_table(users_setting_record) if profile
+  end
+
+  def sync_users_settings_table
+    users_setting_record = Users::Setting.find_by(user_id: id)
+    users_setting_record ||= Users::Setting.create(user_id: id)
```
Comment (rhymes) [inline app/models/user.rb:652]: this can be simplified with [create_or_find_by](https://api.rubyonrails.org/classes/ActiveRecord/Relation.html#method-i-create_or_find_by):

```suggestion
    users_setting_record = Users::Setting.create_or_find_by(user_id: id)
```

File: app/models/user.rb [hunk 7]
```diff
@@ -575,6 +628,45 @@ def flipper_id
...
+
+    migrate_users_and_profile_fields_to_users_settings(users_setting_record)
+  end
+
+  def sync_users_notification_settings_table
+    users_notification_setting_record = Users::NotificationSetting.find_by(user_id: id)
+    users_notification_setting_record ||= Users::NotificationSetting.create(user_id: id)
```
Comment (rhymes) [inline app/models/user.rb:659]: this can be simplified with [create_or_find_by](https://api.rubyonrails.org/classes/ActiveRecord/Relation.html#method-i-create_or_find_by):

```suggestion
    users_notification_setting_record = Users::NotificationSetting.create_or_find_by(user_id: id)
```

File: db/schema.rb (hunk 1 above)
Comment (rhymes) [reply db/schema.rb:1456]: I think so, but the PR is missing the actual migration file

**Review by rhymes — COMMENTED** [review]
Good stuff, it's working!

I left some inline notes and I can add a suggestion to those notes:

Right now each time a field is found in the `.each` loops, a `setting.update()` is issued. This means that if 20 fields have to be udpated, we're sending 20 `UPDATE` statements to the database. Those can be reduced with something like:

```ruby
  def sync_relevant_profile_fields_to_user_settings_table(users_setting_record)
    PROFILE_FIELDS_TO_MIGRATE_TO_USERS_SETTINGS_TABLE.each do |field|
      users_setting_record.assign_attributes(field => profile.data[field]) if profile.data[field].present?
    end
  end

  def migrate_users_and_profile_fields_to_users_settings(users_setting_record)
    USER_FIELDS_TO_MIGRATE_TO_USERS_SETTINGS_TABLE.each do |field|
      if USER_SETTINGS_ENUM_FIELDS.include?(field)
        field_enums = Users::Setting.defined_enums[field]
        users_setting_record.assign_attributes(field => field_enums[public_send(field).to_sym])
      else
        users_setting_record.assign_attributes(field => public_send(field))
      end
    end

    sync_relevant_profile_fields_to_user_settings_table(users_setting_record) if profile

    users_setting_record.save
  end
```

same for the notifications. Basically we assign the attributes during the loops, with `.assign_attributes` and then call `.save` only once at the end. This will create only one `UPDATE`

File: db/schema.rb (hunk 1 above)
Comment (msarit) [reply db/schema.rb:1456]: Note: [The DUS PR](https://github.com/forem/forem/pull/13390) will go in first, which will resolve this apparent change

**Comment by Ridhwana** [comment]
@msarit thanks for updating this PR based on feedback :) It's looking great, what is there still left to do in order to move this into a ready for review state?

**Comment by Ridhwana** [comment]
Quick one: Just noticed that when I run `rake db:seed` I get the ffg: 

```
forem|AritRidhwana/rfc#32-p3-user-settings-sync-code⚡ ⇒ rake db:seed
Seeding with multiplication factor: 1

  1. Organizations already exist. Skipping.
  2. Creating DEV profile fields.
  3. Users already exist. Skipping.
  User with email = admin@forem.local found, skipping.
  4. Creating Tags.
  5. Creating 25 Articles.
rake aborted!
NoMethodError: undefined method `brand_color1' for nil:NilClass
/Users/ridhwana/Projects/DEV/forem/app/models/user.rb:633:in `public_send'
/Users/ridhwana/Projects/DEV/forem/app/models/user.rb:633:in `block in sync_relevant_profile_fields_to_user_settings_table'
/Users/ridhwana/Projects/DEV/forem/app/models/user.rb:632:in `sync_relevant_profile_fields_to_user_settings_table'
/Users/ridhwana/Projects/DEV/forem/app/models/user.rb:647:in `migrate_users_and_profile_fields_to_users_settings'
/Users/ridhwana/Projects/DEV/forem/app/models/user.rb:655:in `sync_users_settings_table'
/Users/ridhwana/Projects/DEV/forem/db/seeds.rb:192:in `block (2 levels) in <main>'
/Users/ridhwana/Projects/DEV/forem/db/seeds.rb:174:in `times'
/Users/ridhwana/Projects/DEV/forem/db/seeds.rb:174:in `block in <main>'
/Users/ridhwana/Projects/DEV/forem/app/lib/seeder.rb:30:in `create_if_none'
/Users/ridhwana/Projects/DEV/forem/db/seeds.rb:173:in `<main>'
Tasks: TOP => db:seed
(See full trace by running task with --trace)
```

**Comment by Ridhwana** [comment]
> brand_color1
```
  def sync_relevant_profile_fields_to_user_settings_table(users_setting_record)
    PROFILE_FIELDS_TO_MIGRATE_TO_USERS_SETTINGS_TABLE.each do |field|
      users_setting_record.assign_attributes(field => profile.public_send(field)) if profile&.public_send(field).present?
    end
  end
```

I think its because profile is `nil` and hence we get `undefined method `brand_color1' for nil:NilClass`. Maybe we should use the safe navigation operator?

File: spec/requests/user/user_settings_spec.rb [hunk 8]
```diff
@@ -238,20 +238,29 @@
       expect(user.reload.profile_updated_at).to be > 2.minutes.ago
     end
 
-    it "disables reaction notifications" do
-      put "/users/#{user.id}", params: { user: { tab: "notifications", reaction_notifications: 0 } }
-      expect(user.reload.reaction_notifications).to be(false)
+    it "disables reaction notifications (in both users and notification_settings tables)" do
```
Comment (msarit) [inline spec/requests/user/user_settings_spec.rb:241]: These are the specs that I added to the `users_setting_spec` file to test the `users_setting` and `users_notification_setting` values being set. Are these sufficient? Please advise.

**Review by msarit — COMMENTED** [review]
Leaving some comments on my changes

File: app/models/users/setting.rb [hunk 9]
```diff
@@ -4,30 +4,19 @@ class Setting < ApplicationRecord
...
+    enum config_navbar: { default: 0, static: 1 }, _suffix: :navbar
+    enum config_theme: { default: 0, minimal_light_theme: 1, night_theme: 2, pink_theme: 3,
                          ten_x_hacker_theme: 4 }
 
     validates :user_id, presence: true
     validates :experience_level, numericality: { less_than_or_equal_to: 10 }, allow_blank: true
     validates :feed_referential_link, inclusion: { in: [true, false] }
     validates :feed_url, length: { maximum: 500 }, allow_nil: true
     validates :inbox_guidelines, length: { maximum: 250 }, allow_nil: true
-    validate :validate_feed_url, if: :feed_url_changed?
```
Comment (msarit) [inline app/models/users/setting.rb:19]: This validation was causing the `Feed::Import` specs to fail vastly and mercilessly.

The `feed_url` attribute is already validated in the User model (`user.rb:735`) and the validation is tested in the associated spec file (`user_spec.rb:280`.

So, Rhymes and I decided that, since we are still in the data-sync stage of the RFC, and the `feed_url` is copied over from the `users` table, that we could remove this validation in `users/setting.rb` for now, and re-add it when we point data updates directly to the `users_settings` table.

File: spec/models/users/setting_spec.rb [hunk 10]
```diff
@@ -4,34 +4,14 @@
...
     let(:user) { create(:user) }
+    let(:setting) { user.setting }
 
     it { is_expected.to validate_length_of(:inbox_guidelines).is_at_most(250).allow_nil }
     it { is_expected.to define_enum_for(:inbox_type).with_values(private: 0, open: 1).with_suffix(:inbox) }
-    it { is_expected.to define_enum_for(:config_font).with_values(default: 0, comic_sans: 1, monospace: 2, open_dyslexic: 3, sans_serif: 4, serif: 5) }
-    it { is_expected.to define_enum_for(:config_navbar).with_values(default_navbar: 0, static_navbar: 1) }
-    it { is_expected.to define_enum_for(:config_theme).with_values(default_theme: 0, minimal_light_theme: 1, night_theme: 2, pink_theme: 3, ten_x_hacker_theme: 4) }
-
-    describe "when validating feed_url", vcr: true do
```
Comment (msarit) [inline spec/models/users/setting_spec.rb:16]: This is the associated spec for the `validate_feed_url` private method that I deleted in `users/setting.rb`

**Comment by msarit** [comment]
Soul Train dancer

@Ridhwana @citizen428 It’s GREEN and ready for your official review! :sweat_smile:

File: app/models/users/setting.rb (hunk 9 above)
Comment (Ridhwana) [reply app/models/users/setting.rb:19]: @msarit I wouldn't advise removing this completely, but rather comment it out with a note. The same goes for the specs. The reason I say that is so that we can remember to put it back in else sometimes things like this slip through the cracks 😅

**Comment by Ridhwana** [comment]
@msarit question - what happens if a user deletes their account whilst we have the sync code, should we check if they have associated users_settings and users_settings notifications and delete those too?

**Comment by rhymes** [comment]
@Ridhwana AFAIK it's fine, the `User` model has:

```ruby
  has_one :notification_setting, class_name: "Users::NotificationSetting", dependent: :destroy
  has_one :setting, class_name: "Users::Setting", dependent: :destroy
```

so both will be destroyed when the account is removed

**Review by rhymes — APPROVED** [review]
Well done!

Let's not forget https://github.com/forem/forem/pull/13390 has to merged first! :)

**Comment by Ridhwana** [comment]
> Well done!
> 
> 
> 
> Let's not forget https://github.com/forem/forem/pull/13390 has to merged first! :) 

@rhymes from what I remember we discussed merging the sync code first then the DUS? https://github.com/forem/forem/pull/13390#issuecomment-834173674

**Comment by Ridhwana** [comment]
> @Ridhwana AFAIK it's fine, the `User` model has:
> 
> 
> 
> ```ruby
> 
>   has_one :notification_setting, class_name: "Users::NotificationSetting", dependent: :destroy
> 
>   has_one :setting, class_name: "Users::Setting", dependent: :destroy
> 
> ```
> 
> 
> 
> so both will be destroyed when the account is removed

🤦🏽‍♀️ I forgot we had this. Thanks !

**Comment by rhymes** [comment]
@Ridhwana https://github.com/forem/forem/pull/13719/files#r632535397 (if this goes first, then migrations in the other PR have to be moved here)

**Comment by Ridhwana** [comment]
> @Ridhwana https://github.com/forem/forem/pull/13719/files#r632535397 (if this goes first, then migrations in the other PR have to be moved here)

Good point. @msarit will you make the call on which PR you want to merge first and then move the migration appropriately please 😊

**Comment by msarit** [comment]
> > @Ridhwana https://github.com/forem/forem/pull/13719/files#r632535397 (if this goes first, then migrations in the other PR have to be moved here)
> 
> Good point. @msarit will you make the call on which PR you want to merge first and then move the migration appropriately please 😊

Thank you @rhymes @Ridhwana I will merge the [Data Update PR](https://github.com/forem/forem/pull/13390) first, before this one. @Ridhwana, you mentioned that Mac could run the DUS on a production clone, for testing purposes. Does that need to happen FIRST before I merge the DUS PR?

**Comment by Ridhwana** [comment]
> Thank you @rhymes @Ridhwana I will merge the [Data Update PR](https://github.com/forem/forem/pull/13390) first, before this one. @Ridhwana, you mentioned that Mac could run the DUS on a production clone, for testing purposes. Does that need to happen FIRST before I merge the DUS PR?

Turns out that the process of cloning the data is going to work out pretty expensive.😞 We chatted more and determined that the risk of the DUS is not as high since we're migrating existing data to new tables rather than destroying record from existing tables. 


The best technique is just to test as much locally as possible. @maestromac suggested possibly testing by creating 10k users locally with some permutations of what we expecting from the data.

File: db/schema.rb (hunk 1 above)
Comment (Zhao-Andy) [reply db/schema.rb:1448]: Hmmmm I think this is fine but am confused as to why it is fine... I guess because the DUS PR has these changes added, but I would think that PR being merged wouldn't show that these lines are newly added in git. 🤔

**Review by Zhao-Andy — APPROVED** [review]
This works for me! Nice job on it :)

**Review by citizen428 — APPROVED** [review]

**Comment by msarit** [comment]
Updating you folks on this PR @Ridhwana @rhymes @citizen428 @andreamartz 😄 

**TL;DR Jamie will re-run the DUS after I deploy this sync-code PR.**

---

Yesterday, Jamie ran the [data-update-scripts](https://github.com/forem/forem/pull/13856) that copied relevant `users_settings` and `users_notification_settings` fields from `users` table.

However, by the time I merge this present sync-code PR, there'll most likely be some user records that do not have entries in `users_settings` and `users_notification_settings`; thereby rendering the sync-code useless for these users.

In hindsight, it made better sense to deploy this sync-code before the data update scripts. 🤓 

I've alerted Jamie to re-run the DUS after I deploy this sync-code 👍🏾

**Comment by msarit** [comment]
@Ridhwana @rhymes @jgaskins

Hello all, I have questions re: the Travis failures on this PR (there are 5 total, I have fixed 1, but have questions on the remaining 4)

**Long story short, these failures are happening for 2 reasons:**
- `Users::Setting` and `Users::NotificationSetting` records are NOW being created together with the spec's `users`, even though the spec tests the creation of these records AFTER the migration runs
- We can no longer set certain attributes to `nil` upon user-creation

I created this short video to describe my proposed fixes, but I'm not sure if this is necessary, seeing as these are specs for migrations that have already run 🤷🏾‍♀️ 
Please advise 🙏🏾  ❤️ 

https://www.loom.com/share/9ecae2c0742f437899e152debfcf0de0

**Comment by Ridhwana** [comment]
> @Ridhwana @rhymes @jgaskins @citizen428
> 
> Hello all, I have questions re: the Travis failures on this PR (there are 5 total, I have fixed 1, but have questions on the remaining 4)
> 
> **Long story short, these failures are happening for 2 reasons:**
> 
> * `Users::Setting` and `Users::NotificationSetting` records are NOW being created together with the spec's `users`, even though the spec tests the creation of these records AFTER the migration runs
> * We can no longer set certain attributes to `nil` upon user-creation
> 
> I created this short video to describe my proposed fixes, but I'm not sure if this is necessary, seeing as these are specs for migrations that have already run 🤷🏾‍♀️
> Please advise 🙏🏾 ❤️
> 
> https://www.loom.com/share/9ecae2c0742f437899e152debfcf0de0

Hi @msarit thank you for the video! It was super helpful to get a quick idea on the problem!

This is a tricky one because like you mention the sync codes add the relevant `users settings` rows when we create the `user` in the `after_commit` callback. 

I was going to suggest that we don't check that the count has _changed_ but rather the final size is 3, but then I realized we wouldn't really be testing what the DUS is doing. 

Hence, I think your approach seems okay because the premise of the test is testing that when there are no related setting rows then the DUS adds them. I would also suggest adding a comment to the code to explain why you're destroying the records, so as to provide context for someone reading it. Keep in mind not only the sync code but also the complete takeover code when you write this comment.

**Comment by rhymes** [comment]
@msarit I agree with @Ridhwana! I think your plan is solid! :-) 
One tiny detail: I would delete only the user's notification setting, not all of them, so `user.notification_setting.destroy` instead of `Users::NotificationSetting.destroy_all`
